### PR TITLE
Plonky3 keccak

### PR DESCRIPTION
**Pull request by qwang98**
Not ready for review yet. Still debugging witness generation. Depends on feature that sets hints in an array in #1645.

Currently merges into #1645 rather than the main. The only changes should be in `keccak.asm` and `mod.asm`.

The Plonky3 Keccak circuit is adapted from the standard Keccak algorithm here: https://keccak.team/keccak_specs_summary.html, which I've proven to be mathematically equivalent. Therefore, Plonky3 Keccak circuit should be sound (has one unique witness for any input).

Our current solver won't be able to populate witness values just given the constraints, which is different from the "sequential" way variables are defined in the standard Keccak algorithm, and that's why hints are needed.

### CODE REVIEW

File: std/machines/hash/keccak.asm [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+    //     pub a_prime_prime_0_0_bits: [T; 64],
+
+    //     /// ```ignore
+    //     /// A'''[0, 0, z] = A''[0, 0, z] ^ RC[k, z]
+    //     /// ```
+    //     pub a_prime_prime_prime_0_0_limbs: [T; U64_LIMBS],
+    // }
+
+    pol commit export;
+    pol commit preimage[100]; // 5 * 5 * 4
```
Comment (chriseth) [inline std/machines/hash/keccak.asm:65]: Why not `pol commit preimage[5 * 5 * 4]`?

File: std/machines/hash/keccak.asm [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(first_step)
+    //                 .assert_eq(local.preimage[y][x][limb], local.a[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::new(100, |i| first_step * (preimage[i] - a[i]) = 0);
```
Comment (chriseth) [inline std/machines/hash/keccak.asm:115]:
```suggestion
    array::zip(preimage, a, |p, a_el| first_step * (p - a_el) = 0);
```

(the naming of a_el is unfortunate, but maybe the name for `a` could also be longer?

File: std/machines/hash/keccak.asm [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    pol constant first_step(i) { if i % NUM_ROUNDS == 0 { 1 } else { 0 } }; // same as step_flags[0]
+    pol constant final_step(i) { if i % NUM_ROUNDS == NUM_ROUNDS - 1 { 1 } else { 0 } }; // same as step_flags[NUM_ROUNDS - 1]
+    pol constant not_final_step(i) { if i % NUM_ROUNDS == NUM_ROUNDS - 1 { 0 } else { 1 } };
```
Comment (chriseth) [inline std/machines/hash/keccak.asm:102]:
```suggestion
    let not_final_step = 1 - final_step;
```
?

File: std/machines/hash/keccak.asm [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+        0x8000000000000080,
+        0x000000000000800A,
+        0x800000008000000A,
+        0x8000000080008081,
+        0x8000000000008080,
+        0x0000000080000001,
+        0x8000000080008008
+    ];
+    
+    let RC_BITS: int[] = [ // 24 * 64
```
Comment (chriseth) [inline std/machines/hash/keccak.asm:412]: Is there a way to compute this from RC?

File: std/machines/hash/keccak.asm (hunk 3 above)
Comment (qwang98) [reply std/machines/hash/keccak.asm:106]: Agree

File: std/machines/hash/keccak.asm (hunk 2 above)
Comment (qwang98) [reply std/machines/hash/keccak.asm:119]: Agree on simplification. Keeping `a` naming for now to be consistent with Plonky3, but will change in the end.

File: std/machines/hash/keccak.asm [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    pol constant first_step(i) { if i % NUM_ROUNDS == 0 { 1 } else { 0 } }; // same as step_flags[0]
```
[the remark this answers is not in the document]
Comment (qwang98) [reply std/machines/hash/keccak.asm:104]: Agree.

File: std/machines/hash/keccak.asm (hunk 1 above)
Comment (qwang98) [reply std/machines/hash/keccak.asm:69]: Agree

**Comment by leonardoalt** [comment]
Closing since done in another PR. Please re-open if I'm wrong.

**Pull request closed without merging**